docs/proposals/04-new-core-gardener-cloud-apis.md

```diff
@@ -0,0 +1,716 @@
+# Optional list of labels on `Seed` resources that marks those seeds whose shoots may use this provider profile.
+# An empty list means that all seeds of the same provider type are supported.
+# This is useful for environments that are of the same type (like openstack) but may have different "instances"/landscapes.
+# seedSelector:
+#   matchLabels:
+#     foo: bar
+  kubernetes:
+    versions:
+    - 1.12.1
+    - 1.11.0
+    - 1.10.5
+  backup: # See https://github.com/gardener/gardener/blob/master/docs/proposals/02-backupinfra.md.
+    type: <some-provider-name> # {aws,azure,gcp,...}
+  # region: eu-west-1
+    secretRef:
+      name: backup-secret
+      namespace: garden
+  machineTypes:
+  - name: m5.large
+    cpu: "2"
+    gpu: "0"
+    memory: 8Gi
+  # storage: 20Gi # optional (not needed in every environment, may only be specified if no volumeTypes have been specified)
+    usable: true
+  volume:
+    minimumSize: 20Gi
+    types: # optional (not needed in every environment, may only be specified if no machineType has a `storage` field)
+    - name: gp2
+      class: standard
+    - name: io1
+      class: premium
+  machineImages:
+  - name: coreos
+    versions:
+    - version: 1967.5.0
+      expirationDate: 2020-04-05T08:00:00Z
+    - version: 2023.5.0
+  - name: ubuntu
+    versions:
+    - version: 18.04.201906170
+  zones: # optional (not needed in every environment)
+  - region: europe-central-1
+    zones:
+    - name: europe-central-1a
+    - name: europe-central-1b
+    - name: europe-central-1c
+    # unavailableMachineTypes: # optional, list of machine types defined above that are not available in this zone
+    # - name: m5.large
+    # unavailableVolumeTypes: # optional, list of volume types defined above that are not available in this zone
+    # - name: io1
+# CA bundle that will be installed onto every shoot machine that is using this provider profile.
+# caBundle: |
+#   -----BEGIN CERTIFICATE-----
+#   ...
+#   -----END CERTIFICATE-----
+  providerConfig:
+    <some-provider-specific-controlplane-config>
+    # We don't have concrete examples for every existing provider yet, but these are the proposals:
+    #
+    # Example for Alicloud:
+    #
+    # apiVersion: alicloud.provider.extensions.gardener.cloud/v1alpha1
+    # kind: ProviderProfileConfig
+    # machineImages:
+    # - name: coreos
+    #   version: 2023.5.0
+    #   id: coreos_2023_4_0_64_30G_alibase_20190319.vhd
+    #
+    #
+    # Example for AWS:
+    #
+    # apiVersion: aws.provider.extensions.gardener.cloud/v1alpha1
+    # kind: ProviderProfileConfig
+    # machineImages:
+    # - name: coreos
+    #   version: 1967.5.0
+    #   regions:
+    #   - name: europe-central-1
+    #     ami: ami-0f46c2ed46d8157aa
+    #
+    #
+    # Example for Azure:
+    #
+    # apiVersion: azure.provider.extensions.gardener.cloud/v1alpha1
+    # kind: ProviderProfileConfig
+    # machineImages:
+    # - name: coreos
+    #   version: 1967.5.0
```

Comment:
The semantic of version should be changed.
Currently, whatever is the version of the image, it must be the same in the `.spec.machineImages` and offere it to the end users. In this way gardener cannot advertise version `a.b.c` as `x.y.z`, but only as `a.b.c`.

So I would propose to replace `publisher`, `offer`, `sku` and `version` with the `urn` from the machine image description. 

While the `version` will contain the version that Gardener wants to advertise. 
So it will looks like 
```yaml
    machineImages:
    - name: coreos
      version: 1967.5.0
      urn: "CoreOS:CoreOS:Stable:2023.4.0"
```

Most probably this will require change in the `provider-azure` to adapt the change.

Note, only for azure the mapping is not possible at the moment:
* for aws a version is mapped to AMIs
* for alicloud the version is mapped to id
* for gcp the version is mapped to image
* for openstack the version is mapped to image

Reply:
I see what you mean, but why do we need to change it to `urn`? Isn't it enough to simply add a separate `version` (ok, `version` cannot be taken because it's already used, but some other similar name) in the Azure machine images spec? Then the mapping you mentioned can be achieved similar to all the other providers.
Changing it to `urn` has an implicit logic with word between the `:` is what, and in the end the Azure provider needs to know what is what because it's passed to MCM (and then to the Azure API). So, let's keep the current structure, if there is no strong reason against it. The separate `version` field can certainly be introduced (actually, it could already be done today and is not really bound to this GEP).

Reply:
Other field for the version was my initial idea, but then it will need new name, that will has the same semantic as `version` of the other providers - I find this confusing, except if we rename the `version` field for the rest of the providers, too.

This is why I proposed to switch to `urn` which has the format `Publisher:Offer:Sku:Version`[1]. And it is usually maintained in the Azure API so that it can be got directly, otherwise it could be  build from the rest of the fields:
```bash
$ az vm image list ...
{                                                                                                                                
    "offer": "CoreOS",                                                                                                             
    "publisher": "CoreOS",                                                                                                         
    "sku": "Stable",                                                                                                               
    "urn": "CoreOS:CoreOS:Stable:2023.5.0",                                                                                        
    "version": "2023.5.0"                                                                                                          
}
```

Take a look in GCP `image` field - it has the format `projects/<project-name>/global/images/<image-name>`, e.g. `image: projects/coreos-cloud/global/images/coreos-stable-2023-5-0-v20190312`. The `urn` field will be pretty much the same. 

We should be able to create VMs just from the URN, checking the docu/help of `az vm  create` - it supports tha flag `--image` that accepts the URN of the desired image
```
$ az vm create --help
....
--image                        : The name of the operating system image as a URN alias, URN,                                  
                                     custom image name or ID, or VHD blob URI. This parameter is                                  
                                     required unless using `--attach-os-disk.` Valid URN format:                                  
                                     "Publisher:Offer:Sku:Version".  Values from: az vm image list,                               
                                     az vm image show.
...
Examples
...
    Create a default RedHat VM with automatic SSH authentication using an image URN.                                              
        az vm create -n MyVm -g MyResourceGroup --image RedHat:RHEL:7-RAW:7.4.2018010506
```




I also checked how the MCM is creating the VMs on azure and azure go sdk. From what I have seen I got the feeling that the azure SDK actually requires the quadruple of `SKU`, `Publisher`, `Offer` and `Version` which has influenced the machine API. 

So I think we have three alternative options:
1. rename `version` to something other (e.g. `providerVersion`, `cloudVersion`, `gardenerVersion`, you name it) for all cloud providers. Gardener will offer to its users the version in the new field and it will be mapped to the `version` in the azure provider
```yaml
  machineImages:
  - name: coreos
    version: 1967.5.0 #  Gardener users will see this version in their azure portal
    publisher: CoreOS
    offer: CoreOS
    sku: Stable
    providerVersion: 2000.1.2 # Gardener users will see this version in the gardener API. This field has to be added to the other providers, too.
```
2. MCM switches to `urn` and parses the `publisher`, `offer`, `sku` and `version` from it.
3. add new field `imageVersion` that will take the semantic of the current `version` field. In this way Gardener will offer to its user the version stored in the `version` field, which on Azure is mapped to `imageVersion`. At the end it will looks like this
```yaml
  machineImages:
  - name: coreos
    imageVersion: 1967.5.0 #  Gardener users will see this version in their azure portal
    publisher: CoreOS
    offer: CoreOS
    sku: Stable
    version: 2000.1.2 # Gardener users will see this version in the gardener API
```

[1]https://docs.microsoft.com/en-us/azure/virtual-machines/linux/cli-ps-findimage

Reply:
I'm fine with switching to `URN` and I think you're right, it makes everything easier.
But, as you said, before we can do that MCM has to support that (currently they [don't](https://github.com/gardener/machine-controller-manager/blob/master/kubernetes/machine_classes/azure-machine-class.yaml#L26-L30)). 
Also, I still think it can and should be done separately and not be part of this GEP (it's more or less a new feature, but with this GEP we only want to migrate existing APIs into new ones, and adding new features in between makes the transition harder). Would you please open an issue at the g/mcm repo and g/gardener-extensions to implement the same after g/mcm is fixed? Thank you.

Reply:
Thanks!

I have opened PRs https://github.com/gardener/machine-controller-manager/pull/326 and https://github.com/gardener/gardener-extensions/pull/290 to address this feature.